File: rag3d-js/src/textproc.js
// Processamento de texto agnóstico de língua — espelha trirag/textproc.py.
//
// Precisa produzir os MESMOS tokens/n-gramas que a versão Python para que os
// hologramas batam entre linguagens. Iteração por code point (Array.from) para
// casar com o fatiamento por code point do Python.

// Faixas CJK (han, ext-A, hiragana, katakana, hangul, compat) — mesmas do Python
const CJK = "[\\u4e00-\\u9fff\\u3400-\\u4dbf\\u3040-\\u309f\\u30a0-\\u30ff\\uac00-\\ud7af\\uf900-\\ufaff]";
const CJK_RE = new RegExp(CJK, "gu");
const CJK_ONE = new RegExp("^" + CJK + "$", "u");
// \w Unicode do Python ~= letras + números + underscore
const WORD_RE = /[\p{L}\p{N}_]+/gu;

// pontuação de fim de sentença multi-escrita (mesma lista do Python).
// . ! ? são literais dentro de [] — sem escape (inválido no modo unicode).
const SENT_END = ".!?。！？؟۔।॥։።၊။…";
const SENT_RE = new RegExp(`[^${SENT_END}\\n]+[${SENT_END}]*\\s*`, "gu");

// sentinela p/ proteger pontos internos de número/seção (13.243, 5.1.2)
const SENT_GUARD = String.fromCharCode(0xe000);

export function normalize(text) {
  text = text.normalize("NFKC");
  // conserta números partidos pela extração de PDF: "13. 243" -> "13.243".
  // [0-9] e (?![0-9]) explícitos (não \d/\b) — \d é ASCII no JS e Unicode no
  // Python; a forma explícita é idêntica nas duas linguagens (mantém paridade).
  text = text.replace(/([0-9])\.\s+([0-9]{3})(?![0-9])/g, "$1.$2");
  text = text.replace(/[ \t]+/g, " ");
  text = text.replace(/\n{3,}/g, "\n\n");
  return text.trim();
}

function codepoints(s) {
  return Array.from(s);
}

export function estimateTokens(text) {
  if (!text) return 0;
  const cjk = (text.match(CJK_RE) || []).length;
  const rest = codepoints(text).length - cjk;
  return cjk + Math.max(1, Math.floor(rest / 4));
}

export function splitSentences(text) {
  // não quebrar sentença no ponto entre dígitos (13.243, 5.1.2, R$ 1.000).
  // [0-9] explícito (não \d) — paridade JS/Python.
  text = text.replace(/(?<=[0-9])\.(?=[0-9])/g, SENT_GUARD);
  const restore = (s) => s.split(SENT_GUARD).join(".");
  const out = [];
  for (const paraRaw of text.split("\n\n")) {
    const para = paraRaw.trim();
    if (!para) continue;
    let matched = false;
    for (const m of para.matchAll(SENT_RE)) {
      const s = m[0].trim();
      if (s) {
        out.push(restore(s));
        matched = true;
      }
    }
    if (!matched && para) out.push(restore(para));
  }
  if (out.length) return out;
  const t = restore(text).trim();
  return t ? [t] : [];
}

// tokens lexicais: palavras minúsculas; CJK vira caracteres + bigramas
export function wordTokens(text, cjkBigrams = true) {
  text = text.toLowerCase();
  const toks = [];
  for (const m of text.matchAll(WORD_RE)) {
    const w = m[0];
    const parts = w.split(CJK_RE);
    const cjks = w.match(CJK_RE) || [];
    for (const p of parts) if (p) toks.push(p);
    if (cjks.length) {
      for (const c of cjks) toks.push(c);
      if (cjkBigrams && cjks.length > 1) {
        for (let i = 0; i + 1 < cjks.length; i++) toks.push(cjks[i] + cjks[i + 1]);
      }
    }
  }
  return toks;
}

// n-gramas de caracteres — base do encoder fallback, qualquer escrita
export function charNgrams(text, nLo = 3, nHi = 5) {
  text = " " + text.toLowerCase().replace(/\s+/g, " ") + " ";
  const cp = codepoints(text);
  const out = [];
  const L = cp.length;
  for (let n = nLo; n <= nHi; n++) {
    for (let i = 0; i + n <= L; i++) out.push(cp.slice(i, i + n).join(""));
  }
  return out;
}

export { CJK_ONE };

File: rag3d-java/src/main/java/io/rag3d/core/TextProc.java
package io.rag3d.core;

import java.text.Normalizer;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Processamento de texto agnóstico de língua — espelha rag3d/textproc.py e
 * rag3d-js/src/textproc.js.
 *
 * <p>Precisa produzir os MESMOS tokens/n-gramas que as outras linguagens para
 * que os hologramas batam. Iteração por CODE POINT (não char UTF-16) para casar
 * com o fatiamento do Python.
 */
public final class TextProc {

    // Faixas CJK (han, ext-A, hiragana, katakana, hangul, compat)
    private static final String CJK =
            "[\\u4e00-\\u9fff\\u3400-\\u4dbf\\u3040-\\u309f\\u30a0-\\u30ff\\uac00-\\ud7af\\uf900-\\ufaff]";
    private static final Pattern CJK_RE = Pattern.compile(CJK);
    private static final Pattern WORD_RE = Pattern.compile("[\\p{L}\\p{N}_]+", Pattern.UNICODE_CHARACTER_CLASS);

    // pontuação de fim de sentença multi-escrita (mesma lista das outras linguagens)
    private static final String SENT_END = ".!?。！？؟۔।॥։።၊။…";
    private static final Pattern SENT_RE =
            Pattern.compile("[^" + Pattern.quote(SENT_END) + "\\n]+[" + Pattern.quote(SENT_END) + "]*\\s*");

    // conserta número partido pela extração de PDF: "13. 243" -> "13.243"
    private static final Pattern NUM_FIX = Pattern.compile("([0-9])\\.\\s+([0-9]{3})(?![0-9])");
    private static final Pattern SPACES = Pattern.compile("[ \\t]+");
    private static final Pattern BLANKS = Pattern.compile("\\n{3,}");

    // sentinela p/ proteger ponto interno de número/seção (13.243, 5.1.2)
    private static final String SENT_GUARD = "\uE000";
    private static final Pattern INNER_DOT = Pattern.compile("(?<=[0-9])\\.(?=[0-9])");

    private TextProc() {}

    public static String normalize(String text) {
        String s = Normalizer.normalize(text, Normalizer.Form.NFKC);
        s = NUM_FIX.matcher(s).replaceAll("$1.$2");
        s = SPACES.matcher(s).replaceAll(" ");
        s = BLANKS.matcher(s).replaceAll("\n\n");
        return s.trim();
    }

    private static int[] codePoints(String s) {
        return s.codePoints().toArray();
    }

    private static String fromCodePoints(int[] cp, int from, int to) {
        return new String(cp, from, to - from);
    }

    public static int estimateTokens(String text) {
        if (text == null || text.isEmpty()) return 0;
        int cjk = 0;
        Matcher m = CJK_RE.matcher(text);
        while (m.find()) cjk++;
        int rest = codePoints(text).length - cjk;
        return cjk + Math.max(1, rest / 4);
    }

    /** Divide em sentenças; não quebra no ponto entre dígitos (13.243, 5.1.2). */
    public static List<String> splitSentences(String text) {
        String guarded = INNER_DOT.matcher(text).replaceAll(SENT_GUARD);
        List<String> out = new ArrayList<>();
        for (String paraRaw : guarded.split("\n\n")) {
            String para = paraRaw.trim();
            if (para.isEmpty()) continue;
            boolean matched = false;
            Matcher m = SENT_RE.matcher(para);
            while (m.find()) {
                String s = m.group().trim();
                if (!s.isEmpty()) {
                    out.add(s.replace(SENT_GUARD, "."));
                    matched = true;
                }
            }
            if (!matched) out.add(para.replace(SENT_GUARD, "."));
        }
        if (!out.isEmpty()) return out;
        String t = guarded.replace(SENT_GUARD, ".").trim();
        if (!t.isEmpty()) out.add(t);
        return out;
    }

    /** Tokens lexicais: palavras minúsculas; CJK vira caracteres + bigramas. */
    public static List<String> wordTokens(String text) {
        String lower = text.toLowerCase();
        List<String> toks = new ArrayList<>();
        Matcher m = WORD_RE.matcher(lower);
        while (m.find()) {
            String w = m.group();
            // partes não-CJK, na ordem (split como Python/JS)
            for (String p : CJK_RE.split(w, -1)) if (!p.isEmpty()) toks.add(p);
            List<String> cjks = new ArrayList<>();
            Matcher cm = CJK_RE.matcher(w);
            while (cm.find()) cjks.add(cm.group());
            if (!cjks.isEmpty()) {
                toks.addAll(cjks);
                for (int i = 0; i + 1 < cjks.size(); i++) toks.add(cjks.get(i) + cjks.get(i + 1));
            }
        }
        return toks;
    }

    /** N-gramas de caracteres (por code point) — base do encoder fallback. */
    public static List<String> charNgrams(String text, int nLo, int nHi) {
        String s = " " + text.toLowerCase().replaceAll("\\s+", " ") + " ";
        int[] cp = codePoints(s);
        List<String> out = new ArrayList<>();
        int L = cp.length;
        for (int n = nLo; n <= nHi; n++) {
            for (int i = 0; i + n <= L; i++) out.add(fromCodePoints(cp, i, i + n));
        }
        return out;
    }

    public static List<String> charNgrams(String text) {
        return charNgrams(text, 3, 5);
    }
}
